File: downloads/python/labs2/ISU/Jewels/IGenerator.py
#!/usr/bin/env python
# coding: UTF-8
#
## @package IGenerator
#
#  IGenerator Interface.
#  Interface representing a utility for generating new icons in a Bejeweled-like video game.
#
#  @author Paulo Roma
#  @since 05/04/2020
#
try:                 # python >= 3.4
    from abc import ABC, ABCMeta, abstractmethod
except ImportError:  # python 2
    from abc import ABCMeta, abstractmethod
    ABC = object


##
# Interface representing a utility for generating new icons
# in a Bejeweled-like video game.
#
class IGenerator(ABC):
    ##
    # Returns a new Icon.
    # @return
    #   a new Icon
    #
    @abstractmethod
    def generate(self): pass

    ##
    # Initializes a given 2D array of Icons with new values.
    # Any existing values in the array are overwritten.
    # @param grid
    #   the 2D array to be initialized
    #
    @abstractmethod
    def initialize(self, grid): pass

File: 3-javascript/jewels/jewels/IGenerator.js
/** @file
 *
 *  Summary.
 *  <p>IGenerator Interface.</p>
 *
 *  Description.
 *  <p>Interface representing a utility for generating new icons in a Bejeweled-like video game.</p>
 *
 *  @author Paulo Roma
 *  @see <a href="/cwdc/3-javascript/jewels/IGenerator.js">source</a>
 *  @since 10/03/2021
 */

/**
 * Interface representing a utility for generating new icons
 * in a Bejeweled-like video game.
 * @interface
 */
export class IGenerator {
  constructor() {
    if (this.constructor === IGenerator) {
      throw new TypeError(
        'Abstract class "IGenerator" cannot be instantiated directly.'
      );
    }
  }

  /** Returns a new Icon.
   *  @return {Icon}
   *    a new Icon
   */
  generate() {
    throw new Error("You have to implement the method generate!");
  }

  /** Initializes a given 2D array of Icons with new values.
   *  Any existing values in the array are overwritten.
   *  @param {Array<Array<Icon>>} grid
   *    the 2D array to be initialized
   */
  initialize(grid) {
    throw new Error("You have to implement the method initialize!");
  }
}
